internal/cli/cli.go

```diff
@@ -0,0 +1,193 @@
+package cli
+
+import (
+	"errors"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"strings"
+
+	log "github.com/sirupsen/logrus"
+	"github.com/spf13/cobra"
+	"github.com/spf13/viper"
+	sopsconf "go.mozilla.org/sops/v3/config"
+	"go.mozilla.org/sops/v3/decrypt"
+)
+
+var SopsNoConfigMatch = errors.New("error loading config: no matching creation rules found")
+
+type RootCommand struct{}
+
+func NewRootCommand() *cobra.Command {
+	root := &RootCommand{}
+	cmd := &cobra.Command{
+		Use:               "sops-precommit",
+		PersistentPreRunE: root.persistentPreRunE,
+		RunE:              root.runE,
+	}
+	cmd.PersistentFlags().String("log-level", "info", "Configure log level")
+	return cmd
+}
+
+type decrypter interface {
+	File(filepath string, ext string) ([]byte, error)
+}
+
+type sopsRuleMatcher interface {
+	IsFileMatchCreationRule(file string) (bool, error)
+	HasConf() bool
+}
+
+type sopsclient struct {
+	ConfPath string
+}
+
+func (s *sopsclient) File(filepath string, ext string) ([]byte, error) {
+	return decrypt.File(filepath, ext)
+}
+
+func (s *sopsclient) IsFileMatchCreationRule(file string) (bool, error) {
+	c, err := sopsconf.LoadCreationRuleForFile(s.ConfPath, file, map[string]*string{})
+	if err != nil && err.Error() == SopsNoConfigMatch.Error() {
+		log.Debugf("File: %s doesn't match any sops config creation_rule regex. Skipping.\n", file)
+		return false, nil
+	}
+
+	if err != nil {
+		return false, err
+	}
+
+	if c != nil {
+		return true, nil
+	}
+	return false, nil
+}
+
+func (s *sopsclient) HasConf() bool {
+	return s.ConfPath != ""
+}
+
+func (c *RootCommand) persistentPreRunE(cmd *cobra.Command, args []string) error {
+	// bind flags to viper
+	viper.SetEnvKeyReplacer(strings.NewReplacer("-", "_"))
+	viper.SetEnvPrefix("sops")
+	viper.AutomaticEnv()
+	if err := viper.BindPFlags(cmd.Flags()); err != nil {
+		return err
+	}
+	// set log level
+	logLevel, err := log.ParseLevel(viper.GetString("log-level"))
+	if err != nil {
+		return err
+	}
+	log.SetLevel(logLevel)
+	return nil
+}
+
+func (c *RootCommand) runE(cmd *cobra.Command, args []string) error {
+	log.Debug("debug logging enabled")
+
+	sops := &sopsclient{}
+
+	// List of files in the change set
+	files := []string{}
+
+	if len(args) < 1 {
+		// Try to parse the change set from a pipe
+		input, err := ioutil.ReadAll(os.Stdin)
```

Comment:
This blocks forever if you don't pass arguments on the command line because os.Stdin will be 0 and not provide an EOF to ReadAll.